com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java

```diff
@@ -0,0 +1,240 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class AbstractProtocolServer {
+    private static final Logger logger = Logger.getLogger("java-debug");
+    private static final int BUFFER_SIZE = 4096;
+    private static final String TWO_CRLF = "\r\n\r\n";
+    private static final Pattern CONTENT_LENGTH_MATCHER = Pattern.compile("Content-Length: (\\d+)");
+    private static final Charset PROTOCOL_ENCODING = StandardCharsets.UTF_8; // vscode protocol uses UTF-8 as encoding format.
+
+    private Reader reader;
+    private Writer writer;
+
+    private ByteBuffer rawData;
+    private boolean terminateSession = false;
+    private int contentLength = -1;
+    private AtomicInteger sequenceNumber = new AtomicInteger(1);
+
+    private boolean isDispatchingData;
+    private ConcurrentLinkedQueue<Messages.Event> eventQueue;
+
+    /**
+     * Constructs a protocol server instance based on the given input stream and output stream.
+     * @param input
+     *              the input stream
+     * @param output
+     *              the output stream
+     */
+    public AbstractProtocolServer(InputStream input, OutputStream output) {
+        this.reader = new BufferedReader(new InputStreamReader(input, PROTOCOL_ENCODING));
+        this.writer = new PrintWriter(new BufferedWriter(new OutputStreamWriter(output, PROTOCOL_ENCODING)));
+        this.contentLength = -1;
+        this.rawData = new ByteBuffer();
+        this.eventQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    /**
+     * A while-loop to parse input data and send output data constantly.
+     */
+    public void start() {
+        char[] buffer = new char[BUFFER_SIZE];
+        try {
+            while (!this.terminateSession) {
+                int read = this.reader.read(buffer, 0, BUFFER_SIZE);
+                if (read == -1) {
+                    break;
+                }
+
+                this.rawData.append(new String(buffer, 0, read).getBytes(PROTOCOL_ENCODING));
+                this.processData();
+            }
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, String.format("Read data from io exception: %s", e.toString()), e);
+        }
+    }
+
+    /**
+     * Sets terminateSession flag to true. And the dispatcher loop will be terminated after current dispatching operation finishes.
+     */
+    public void stop() {
+        this.terminateSession = true;
+    }
+
+    public boolean isTerminating() {
```

Comment:
Quite curious about the different about 
```isTerminating``` and "terminateSession```

Reply:
duplicate. Remove `isTerminating` method.